src/store/swapOrderSlice.ts

```diff
@@ -0,0 +1,197 @@
+import type { listenerMiddlewareInstance } from '@/store'
+import { createSelector, createSlice } from '@reduxjs/toolkit'
+import type { OrderStatuses } from '@safe-global/safe-gateway-typescript-sdk'
+import type { RootState } from '@/store'
+import { isSwapTxInfo, isTransactionListItem } from '@/utils/transaction-guards'
+import { txHistorySlice } from '@/store/txHistorySlice'
+import { showNotification } from '@/store/notificationsSlice'
+import { selectSafeInfo } from '@/store/safeInfoSlice'
+
+type AllStatuses = OrderStatuses | 'created'
+type Order = {
+  orderUid: string
+  status: AllStatuses
+  txId?: string
+}
+
+type SwapOrderState = {
+  [orderUid: string]: Order
+}
+
+const initialState: SwapOrderState = {}
+
+const slice = createSlice({
+  name: 'swapOrders',
+  initialState,
+  reducers: {
+    setSwapOrder: (state, { payload }: { payload: Order }): SwapOrderState => {
+      return {
+        ...state,
+        [payload.orderUid]: {
+          ...state[payload.orderUid],
+          ...payload,
+        },
+      }
+    },
+  },
+})
+
+export const { setSwapOrder } = slice.actions
+const selector = (state: RootState) => state[slice.name]
+export const swapOrderSlice = slice
+export const selectAllSwapOrderStatuses = selector
+
+export const selectSwapOrderStatus = createSelector(
+  [selectAllSwapOrderStatuses, (_, uid: string) => uid],
+  (allOrders, uid): undefined | AllStatuses => {
+    return allOrders ? allOrders[uid]?.status : undefined
+  },
+)
+
+const groupKey = 'swap-order-status'
+/**
+ * Listen for changes in the swap order status and determines if a notification should be shown
+ *
+ * Some gotchas:
+ * If the status of an order is created, presignaturePending, open - we always display a notification.
+ * Here it doesn't matter if the order was started through the UI or the gateway returned that order on a new browser instance.
+ *
+ * For fulfilled, expired, cancelled - we only display a notification if the old status is not undefined.
+ * Why? Because if the status is undefined, it means that the order was just fetched from the gateway, and
+ * it was already processed and there is no need to show a notification. If the status is != undefined, it means
+ * that the user has started the swap through the UI (or has continued it from a previous state), and we should show a notification.
+ *
+ * @param listenerMiddleware
+ */
+export const swapOrderStatusListener = (listenerMiddleware: typeof listenerMiddlewareInstance) => {
+  listenerMiddleware.startListening({
+    actionCreator: slice.actions.setSwapOrder,
+    effect: (action, listenerApi) => {
+      const { dispatch } = listenerApi
+      const swapOrder = action.payload
+      const oldStatus = selectSwapOrderStatus(listenerApi.getOriginalState(), swapOrder.orderUid)
+      const newStatus = swapOrder.status
+
+      if (oldStatus === newStatus) {
+        return
+      }
```

Comment:
How can this case happen like calling `setSwapOrder` twice with the same status?

Reply:
Easy. The state is written to from the txHistory and from the widget. The widget doesn't know anything about our state, it has it's own state. And if you leave the widget when the status is open and return with a fullfilled state, then the widget will notify us and we'll try to update the order status. But if the status was updated by txHistory before that... it won't be necessary to show any notification